src/frontend/context/SecurityContext.tsx

```diff
@@ -0,0 +1,130 @@
+import * as React from "react";
+import AsyncStorage from "@react-native-async-storage/async-storage";
+
+import { PASSWORD_KEY } from "../constants";
+
+type AppModeTypes = "normal" | "kill";
+type AuthStatusTypes = "pending" | "authenticated" | "notRequired";
+
+type AuthState = {
+  passcode: string | null;
+  appMode: AppModeTypes;
+  killModeEnabled: boolean;
+  authStatus: AuthStatusTypes;
+};
+
+type KillModeActions =
+  | { type: "setPasscode"; newPasscode: string | null }
+  | { type: "toggleAppMode"; newAppMode?: AppModeTypes }
+  | { type: "toggleKillModeEnabled"; newKillModeValue?: boolean }
+  | { type: "setAuthStatus"; newAuthStatus: AuthStatusTypes };
+
+function killModeReducer(state: AuthState, action: KillModeActions): AuthState {
+  switch (action.type) {
+    /***
+     * if passcode is set to null we need to set the auth status to 'notRequired'
+     *
+     * If a passcode is set, we want to set the authStatus to 'authenticated'
+     * If the auth status is set to 'pending' they will be navigated to the auth screen, to type in their password.
+     */
+    case "setPasscode":
+      const passcode = action.newPasscode;
+      if (passcode === null || validPasscode(passcode)) {
+        AsyncStorage.setItem(PASSWORD_KEY, passcode || "");
+        if (!passcode) {
+          return { ...state, passcode, authStatus: "notRequired" };
+        }
+        return { ...state, passcode, authStatus: "authenticated" };
+      }
+      throw new Error("Invalid New Password");
```

Comment:
would move this to the beginning of the function and write the appropriate check to see if we should throw or not. something along the lines of

```ts
const passcodeIsValid = passcode === null || validPassCode(passcode);

if (!passcodeIsValid) throw new Error("Invalid New Password")

// state update logic here